### PR TITLE
enhance various dynamic-related functions

### PR DESCRIPTION
**Pull request by mr-tz**
There are mostly inspired from looking at a comparison from a fabricated CAPE JSON result and a vivisect capa extraction.
Plus some other things I came across. Happy to discuss anything and just take the best/desired functionality from this PR.

updated show-features output for thread-level features:
2023-07-06_16-14-18_WindowsTerminal

### Checklist

- [x] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/features/extractors/cape/global_.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def extract_features(static) -> Iterator[Tuple[Feature, Address]]:
 
 
 GLOBAL_HANDLER = (
-    extract_arch,
     extract_format,
     extract_os,
+    extract_arch,
```
Comment (mr-tz) [inline capa/features/extractors/cape/global_.py:93]: so we're consistent with the order of the vivisect extractor

File: capa/features/extractors/cape/thread.py [hunk 2]
```diff
@@ -40,14 +40,14 @@ def extract_call_features(behavior: Dict, ph: ProcessHandle, th: ThreadHandle) -
...
+        caller = DynamicAddress(call["id"], int(call["caller"], 16))
         yield API(call["api"]), caller
-        for arg in call["arguments"]:
+        for n, arg in enumerate(call["arguments"]):
             try:
-                yield Number(int(arg["value"], 16)), caller
+                yield Number(int(arg["value"], 16), description=f"arg{n}"), caller
             except ValueError:
-                yield String(arg["value"]), caller
+                yield String(arg["value"], description=f"arg{n}"), caller
```
Comment (mr-tz) [inline capa/features/extractors/cape/thread.py:50]: return_address doesn't really make sense here, but it's the best we can do right now to tie this back to a VA

File: capa/features/extractors/cape/thread.py [hunk 3]
```diff
@@ -40,14 +40,14 @@ def extract_call_features(behavior: Dict, ph: ProcessHandle, th: ThreadHandle) -
...
-        caller = int(call["caller"], 16)
-        caller = AbsoluteVirtualAddress(caller)
+        # TODO this address may vary from the PE header, may read actual base from procdump.pe.imagebase or similar
+        caller = DynamicAddress(call["id"], int(call["caller"], 16))
         yield API(call["api"]), caller
-        for arg in call["arguments"]:
+        for n, arg in enumerate(call["arguments"]):
             try:
-                yield Number(int(arg["value"], 16)), caller
+                yield Number(int(arg["value"], 16), description=f"arg{n}"), caller
```
Comment (mr-tz) [inline capa/features/extractors/cape/thread.py:48]: `arg=<n>` is a crutch here to help interpret results in the show-features output

**Comment by mr-tz** [comment]
updated:
2023-07-06_21-24-40_WindowsTerminal

**Review by williballenthin — APPROVED** [review]

File: capa/features/address.py [hunk 4]
```diff
@@ -36,6 +36,28 @@ def __hash__(self):
...
 
 
+class DynamicAddress(Address):
+    """an address from a dynamic analysis trace"""
+
+    def __init__(self, id_: int, return_address: int):
+        assert id_ >= 0
+        assert return_address >= 0
+        self.id = id_
+        self.return_address = return_address
```
Comment (yelhamer) [inline capa/features/address.py:46]: the _id_ field should make it easier to extract api sequences in the future. I am not sure about the name of the return_address field however, isn't that address of the call instruction (the one before the return address?), not sure if that's a big change anyways.

File: capa/features/extractors/cape/file.py [hunk 5]
```diff
@@ -35,9 +35,34 @@ def rec(process):
...
+        "imports": {
+            "ADVAPI32": {
+                "dll": "ADVAPI32.dll",
+                "imports": [{"address": "0x522000", "name": "OpenSCManagerA"}, ...],
+                ...
+            },
+            ...
+        }
+    """
+    if isinstance(imports, dict):
```
Comment (yelhamer) [inline capa/features/extractors/cape/file.py:62]: nice :)

**Review by yelhamer — COMMENTED** [review]
Thanks @mr-tz ! nice additions :)

one thing -- I suggest renaming _DynamicAddress.return_address_ to maybe _calling_address_ or _caller_? thoughts?

File: capa/features/address.py (hunk 4 above)
Comment (mr-tz) [reply capa/features/address.py:46]: While it's called `caller`, the value is actually the location returned to. I see that `return_address` can be confusing though...
2023-07-07_12-11-23_ida64

If you think we should stick to `caller`, I'm fine with that.

Comment (yelhamer) [reply capa/features/address.py:46]: ah I see, in that case I think we should stick to _return_address_ then.

**Review by yelhamer — APPROVED** [review]

File: capa/features/extractors/cape/thread.py [hunk 6]
```diff
@@ -40,14 +40,15 @@ def extract_call_features(behavior: Dict, ph: ProcessHandle, th: ThreadHandle) -
         if call["thread_id"] != tid:
             continue
 
-        caller = int(call["caller"], 16)
-        caller = AbsoluteVirtualAddress(caller)
-        yield API(call["api"]), caller
-        for arg in call["arguments"]:
+        # TODO this address may vary from the PE header, may read actual base from procdump.pe.imagebase or similar
+        caller = DynamicAddress(call["id"], int(call["caller"], 16))
+        # list similar to disassembly: arguments left-to-right, call
```
Comment (mr-tz) [inline capa/features/extractors/cape/thread.py:45]:
```suggestion
        # list similar to disassembly: arguments right-to-left, call
```

**Pull request merged**